[Build] v1.20.0 min GCC version can't build on armv8

### Describe the issue

The GCC version is checked to be at least version 9 in the `CMakeLists.txt`. https://github.com/microsoft/onnxruntime/blob/c4fb724e810bb496165b9015c77f402727392933/cmake/CMakeLists.txt#L69

When building on armv8 with GCC 9 the check further down fails https://github.com/microsoft/onnxruntime/blob/c4fb724e810bb496165b9015c77f402727392933/cmake/CMakeLists.txt#L693

It seems the feature modifier `bf16` was introduced in GCC 10 [see here](https://gcc.gnu.org/onlinedocs/gcc-10.5.0/gcc/AArch64-Options.html#aarch64-feature-modifiers).  
If I'm not mistaken the minimum GCC version checked for should be rather be 10 instead of 9.

### Urgency

_No response_

### Target platform

armv8

### Build script

Using [`build.py`](https://github.com/microsoft/onnxruntime/blob/c4fb724e810bb496165b9015c77f402727392933/tools/ci_build/build.py) with ` --parallel 4 --config Release --build_dir build --cmake_extra_defines FETCHCONTENT_TRY_FIND_PACKAGE_MODE=NEVER`

### Error / output

```
-- Performing Test HAS_ARM64_BFLOAT16 - Failed
CMake Error at CMakeLists.txt:695 (message):
  The compiler doesn't support BFLOAT16!!!
```

### Visual Studio Version

_No response_

### GCC / Compiler Version

9.5.0

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -673,20 +673,19 @@ else()
   endif()
   check_function_exists(reallocarray HAS_REALLOCARRAY)
   if (NOT APPLE AND NOT CMAKE_SYSTEM_NAME STREQUAL "Emscripten" AND onnxruntime_target_platform STREQUAL "aarch64")
-   # GCC < 12 on aarch64 ships with an assembler that does not accept the
-   # `+bf16` / `+fp16` architecture extensions used below, producing a cryptic
-   # "compiler doesn't support BFLOAT16" failure during the capability probe.
-   # See microsoft/onnxruntime#22837.
-   if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_LESS 12)
-     message(FATAL_ERROR "GCC >= 12 is required for aarch64 BFloat16/Float16 support. Found GCC ${CMAKE_C_COMPILER_VERSION}.")
-   endif()
    check_cxx_compiler_flag(-march=armv8.2-a+bf16 HAS_ARM64_BFLOAT16)
    if(NOT HAS_ARM64_BFLOAT16)
-     message(FATAL_ERROR  "The compiler doesn't support BFLOAT16!!!")
+     message(FATAL_ERROR "The compiler/assembler does not support the aarch64 +bf16 ISA extension. "
+       "On aarch64 this typically requires GCC >= 12 paired with a sufficiently recent binutils "
+       "(the capability is gated by the assembler, not the compiler alone). "
+       "See https://github.com/microsoft/onnxruntime/issues/22837 for details.")
    endif()
    check_cxx_compiler_flag(-march=armv8.2-a+fp16 HAS_ARM64_FLOAT16)
    if(NOT HAS_ARM64_FLOAT16)
-     message(FATAL_ERROR  "The compiler doesn't support FLOAT16!!!")
+     message(FATAL_ERROR "The compiler/assembler does not support the aarch64 +fp16 ISA extension. "
+       "On aarch64 this typically requires GCC >= 12 paired with a sufficiently recent binutils "
+       "(the capability is gated by the assembler, not the compiler alone). "
+       "See https://github.com/microsoft/onnxruntime/issues/22837 for details.")
    endif()
   endif()
 

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -673,6 +673,13 @@ else()
   endif()
   check_function_exists(reallocarray HAS_REALLOCARRAY)
   if (NOT APPLE AND NOT CMAKE_SYSTEM_NAME STREQUAL "Emscripten" AND onnxruntime_target_platform STREQUAL "aarch64")
+   # GCC < 12 on aarch64 ships with an assembler that does not accept the
+   # `+bf16` / `+fp16` architecture extensions used below, producing a cryptic
+   # "compiler doesn't support BFLOAT16" failure during the capability probe.
+   # See microsoft/onnxruntime#22837.
+   if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_LESS 12)
+     message(FATAL_ERROR "GCC >= 12 is required for aarch64 BFloat16/Float16 support. Found GCC ${CMAKE_C_COMPILER_VERSION}.")
+   endif()
    check_cxx_compiler_flag(-march=armv8.2-a+bf16 HAS_ARM64_BFLOAT16)
    if(NOT HAS_ARM64_BFLOAT16)
      message(FATAL_ERROR  "The compiler doesn't support BFLOAT16!!!")

